### PR TITLE
SSCS-10651 Convert Azure Queue to JMS

### PR DESCRIPTION
**Pull request by MohammedHMCTS**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10493

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        jmsConnectionFactory.setClientID(clientId);
+        return jmsConnectionFactory;
+    }
+
+    @Bean
+    @ConditionalOnProperty("flags.tribunals-to-hearings-api.enabled")
+    public ConnectionFactory tribunalsHearingsJmsConnectionFactory(@Value("${spring.application.name}") final String clientId) {
+        String connection = String.format("amqps://%1s?amqp.idleTimeout=%2d", connectionString, idleTimeout);
+        log.info(connection);
+        return new CachingConnectionFactory(jmsConnectionFactory(connection, clientId));
```
[the remark this answers is not in the document]
Comment (dattipoeHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:52]: And the log message shouldn't be there

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    )
+    public void handleIncomingMessage(HearingRequest message) throws TribunalsEventProcessingException {
+        log.info("Message Received");
+        String caseId = null;
+        try {
+            caseId = message.getCcdCaseId();
+            HearingState event = message.getHearingState();
+            log.info("Attempting to process hearing event {} from hearings event queue for case ID {}",
+                     event, caseId
+            );
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java:39]: Weird formatting

File: src/main/java/uk/gov/hmcts/reform/sscs/model/TribunalsDeadLetterMessage.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.hmcts.reform.sscs.model.hearings.HearingRequest;
+
+@Builder(toBuilder = true)
+@AllArgsConstructor
+@Data
+public class TribunalsDeadLetterMessage implements Message {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/model/TribunalsDeadLetterMessage.java:11]: Unsure if this is needed

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]
Nice one, nearly there

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @Value("${azure.service-bus.tribunals-to-hearings-api.receiveTimeout}")
+    private Long receiveTimeout;
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.idleTimeout}")
+    private Long idleTimeout;
+
+    private JmsConnectionFactory jmsConnectionFactory(String connection, final String clientId) {
+        JmsConnectionFactory jmsConnectionFactory = new JmsConnectionFactory(connection);
+        jmsConnectionFactory.setUsername(username);
+        jmsConnectionFactory.setRedeliveryPolicy(new TribunalsHearingsRedeliveryPolicy());
```
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:41]: Good call it has been removed.

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+            hearingsService.processHearingRequest(message);
+            log.info("Hearing event {} for case ID {} successfully processed", event, caseId);
+        } catch (GetCaseException | UnhandleableHearingStateException | UpdateCaseException
+                 | InvalidMappingException ex) {
+            log.error("An exception occurred whilst processing hearing event for case ID {}."
+                          + " Abandoning message", caseId, ex);
+            throw new TribunalsEventProcessingException("Abandoned Message", ex);
```
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java:47]: Good shout message moved to exception

File: src/main/java/uk/gov/hmcts/reform/sscs/model/TribunalsDeadLetterMessage.java (hunk 3 above)
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/model/TribunalsDeadLetterMessage.java:11]: Correct no longer needed

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:52]: I separated them due to it being one method one purpose but I have merged them now and removed logger message.

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -78,11 +78,13 @@ azure:
...
-      connectionString: ${SSCS_SERVICEBUS_CONNECTION_STRING:}
+      connectionString: ${TRIBUNALS_EVENT_QUEUE_CONNECTION_STRING:}
       queueName: ${TRIBUNALS_EVENT_QUEUE_NAME:}
-      retryTimeout: ${TRIBUNALS_EVENT_QUEUE_RETRY_TIMEOUT:1}
-      maxRetries: ${TRIBUNALS_EVENT_QUEUE_RETRIES:5}
-      retryDelay: ${TRIBUNALS_EVENT_QUEUE_RETRY_DELAY:1}
+      receiveTimeout: ${TRIBUNALS_EVENT_QUEUE_RECIVE_TIMEOUT:1}
+      idleTimeout: ${TRIBUNALS_EVENT_QUEUE_IDLE_TIMEOUT:3600000}
+      maxRedeliveries: ${TRIBUNALS_EVENT_QUEUE_MAX_REDELIVERIES:1}
+      username: ${TRIBUNALS_EVENT_QUEUE_POLICY_NAME:}
```
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply src/main/resources/application.yaml:86]: Added

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/CustomMessageConverter.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.io.IOException;
+import javax.jms.BytesMessage;
+import javax.jms.JMSException;
+import javax.jms.Session;
+
+@Component
+public class CustomMessageConverter extends MappingJackson2MessageConverter {
+
+    private static final String TYPE_ID_PROPERTY = "_type";
```
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/CustomMessageConverter.java]: Yes correct I have converted the message on Tribunals will go though testing if not I will switch it to how you have done it.

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    public TribunalsHearingsCustomMessageConverter() {
+        super();
+        this.setTargetType(MessageType.BYTES);
+        this.setTypeIdPropertyName(TYPE_ID_PROPERTY);
+    }
+
+    @NotNull
+    @Override
+    protected BytesMessage mapToBytesMessage(@NotNull Object object, @NotNull Session session, @NotNull ObjectWriter objectWriter)
```
Comment (wmarshid) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java:31]: nice seeing use of @notnull 👌

**Review by wmarshid — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.jms.support.converter.MessageType;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import javax.jms.BytesMessage;
+import javax.jms.JMSException;
+import javax.jms.Session;
+
+@Component
+public class TribunalsHearingsCustomMessageConverter extends MappingJackson2MessageConverter {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java:18]: You can call this something like JsonMessage converter, nothing but the name ties it to Tribunals Hearings

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    @Value("${azure.service-bus.tribunals-to-hearings-api.password}")
+    private String password;
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.receiveTimeout}")
+    private Long receiveTimeout;
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.idleTimeout}")
+    private Long idleTimeout;
+
+    private final String connection = String.format("amqps://%1s?amqp.idleTimeout=%2d", connectionString, idleTimeout);
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:37]: I'd make the template a static final string then call the format in the factory method.

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Bean
+    @ConditionalOnProperty("flags.tribunals-to-hearings-api.enabled")
+    public JmsListenerContainerFactory<DefaultMessageListenerContainer> tribunalsHearingsEventQueueContainerFactory(
+        ConnectionFactory tribunalsHearingsJmsConnectionFactory,
+        DefaultJmsListenerContainerFactoryConfigurer defaultJmsListenerContainerFactoryConfigurer) {
+        DefaultJmsListenerContainerFactory factory = new DefaultJmsListenerContainerFactory();
+        factory.setConnectionFactory(tribunalsHearingsJmsConnectionFactory);
+        factory.setReceiveTimeout(receiveTimeout);
+        factory.setSubscriptionDurable(Boolean.TRUE);
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:57]: Don't think the durable flag is needed for a queue

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        this.hearingsService = hearingsService;
+    }
+
+    @JmsListener(
+        destination = "${azure.service-bus.tribunals-to-hearings-api.queueName}",
+        containerFactory = "tribunalsHearingsEventQueueContainerFactory"
+    )
+    public void handleIncomingMessage(HearingRequest message) throws TribunalsEventProcessingException {
+        log.info("Message Received");
+        String caseId;
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java:33]: Why initialise out here?

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]

**Comment by dattipoeHMCTS** [comment]
Slight concern about the behaviour with session enabled queues...we'll have to keep an eye on it

File: build.gradle [hunk 13]
```diff
@@ -361,6 +361,8 @@ dependencies {
   implementation group: 'com.azure', name: 'azure-core', version: versions.azureCore
   implementation group: 'com.azure', name: 'azure-messaging-servicebus', version: versions.azureMessagingServicebus
 
+  implementation group: 'org.apache.qpid', name: 'qpid-jms-client', version: versions.qpidJmsClient
```
[the remark this answers is not in the document]
Comment (MohammedHMCTS) [reply build.gradle:364]: Ah yeah interesting there was two maybe it didn't exist when the PR was made. Removed one now

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java (hunk 9 above)
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsCustomMessageConverter.java:18]: Good shout did have a generic name before but renamed it based on it only being used within my queue but have updated now to Generic name

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java (hunk 11 above)
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:57]: Correct subscription duration not required unsure why I added it back it was removed at one point...

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java (hunk 12 above)
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java:33]: Was previous used within catch but have moved it into the try

File: build.gradle [hunk 14]
```diff
@@ -361,6 +360,8 @@ dependencies {
...
```
Comment (dattipoeHMCTS) [inline build.gradle:363]: (Minor) makes more sense grouped with the other apache dependencies

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.password}")
+    private String password;
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.receiveTimeout}")
+    private Long receiveTimeout;
+
+    @Value("${azure.service-bus.tribunals-to-hearings-api.idleTimeout}")
+    private Long idleTimeout;
+    public static final String AMQP_CONNECTION_STRING_TEMPLATE = "amqps://%1s?amqp.idleTimeout=%2d";
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:36]: Formatting

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    @Bean
+    @ConditionalOnProperty("flags.tribunals-to-hearings-api.enabled")
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:38]: (Minor) I don't know why I put the ConditionalOnProperty separately on mine but you can just have one at class level

Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/TribunalsHearingsJmsConfig.java:38]: Yeah I had it at the top originally but assume you done it for a reason have updated it now!

File: build.gradle (hunk 14 above)
Comment (MohammedHMCTS) [reply build.gradle:363]: Yeah good shout

File: src/test/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventTopicListenerTest.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+        verify(hearingsService, times(1)).processHearingRequest((eq(hearingRequest)));
+    }
+
+    @Test
+    @DisplayName("When an invalid request comes in make sure exception is thrown")
+    void whenAnInvalidRequestComesIn_makeSureExceptionIsThrown() throws Exception {
+
+        HearingRequest hearingRequest = new HearingRequest();
+
+        doThrow(GetCaseException.class).when(hearingsService).processHearingRequest(eq(hearingRequest));
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventTopicListenerTest.java:52]: Could add more tests for the different exceptions (parameterised)

**Review by ss-evoco — DISMISSED** [review]

**Review by Jonathan-Weston1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/JsonMessageConvertor.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
+package uk.gov.hmcts.reform.sscs.config.jms;
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/jms/JsonMessageConvertor.java:1]: this is a service, not a config, can you move it?

Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/JsonMessageConvertor.java:1]: maybe to into sscs.jms?

File: src/test/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventTopicListenerTest.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+    private static Stream<Arguments> throwableParameters() {
+        return Stream.of(
+            Arguments.of(GetCaseException.class),
+            Arguments.of(UnhandleableHearingStateException.class),
+            Arguments.of(UpdateCaseException.class),
+            Arguments.of(InvalidMappingException.class),
+            Arguments.of(HearingChannelNotFoundException.class)
+        );
+    }
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventTopicListenerTest.java:75]: This test seems a bit pointless? Just try one?

Comment (MohammedHMCTS) [reply src/test/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventTopicListenerTest.java:75]: I was only testing the one at the start but was told to test them all by @dattipoeHMCTS can remove if this is pointless? Please advice and will update.

File: src/main/java/uk/gov/hmcts/reform/sscs/config/jms/JsonMessageConvertor.java (hunk 18 above)
Comment (MohammedHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/config/jms/JsonMessageConvertor.java:1]: Ah you are right I will move it to JMS file.

**Review by Jonathan-Weston1 — APPROVED** [review]

**Review by ss-evoco — APPROVED** [review]

**Pull request merged**